### PR TITLE
feat(cfd): initial CFD stub and tests

### PR DESCRIPTION
**Pull request by chrismaree**
**Motivation**

This PR adds initial CFD smart contract and unit tests. A number of basic libraries were also added.

### CODE REVIEW

File: packages/core/contracts/financial-templates/common/financial-product-libraries/ContractForDifferenceFinancialProductLibrary.sol [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+pragma solidity ^0.8.0;
+import "../../../common/implementation/FixedPoint.sol";
+
+abstract contract ContractForDifferenceFinancialProductLibrary {
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/common/financial-product-libraries/ContractForDifferenceFinancialProductLibrary.sol:5]: Why is this abstract if the method is implemented? Should this be an interface?

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+    modifier onlyOpenState() {
+        require(contractState == ContractState.Open, "Contract state is not OPEN");
+        _;
+    }
+
+    constructor(
+        uint32 _expirationTimestamp,
+        uint256 _collateralPerPair,
+        address _longTokenAddress,
+        address _shortTokenAddress,
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:73]: nit: can we just make these parameter types `ExpandedIERC20` for simplicity?

Same suggestion for the other address params here.

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+
+        collateralToken.safeTransferFrom(msg.sender, address(this), collateralUsed);
+
+        require(longToken.mint(msg.sender, tokensToMint));
+        require(shortToken.mint(msg.sender, tokensToMint));
+    }
+
+    function redeem(uint256 tokensToRedeem) public preExpiration() nonReentrant() returns (uint256 collateralReturned) {
+        require(longToken.burnFrom(msg.sender, tokensToRedeem));
+        require(shortToken.burnFrom(msg.sender, tokensToRedeem));
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:106]: This is so nice.

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+        collateralToken.safeTransferFrom(msg.sender, address(this), collateralReturned);
+    }
+
+    function expire() public postExpiration() onlyOpenState() nonReentrant() {
+        _requestOraclePriceExpiration();
+        contractState = ContractState.ExpiredPriceRequested;
+    }
+
+    function settleExpired(uint256 longTokensToRedeem, uint256 shortTokensToRedeem)
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:118]: Thoughts on naming this settle?

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+        longToken = ExpandedIERC20(_longTokenAddress);
+        shortToken = ExpandedIERC20(_shortTokenAddress);
+        collateralToken = IERC20(_collateralAddress);
+        priceIdentifier = _priceIdentifier;
+
+        financialProductLibrary = ContractForDifferenceFinancialProductLibrary(_financialProductLibrary);
+    }
+
+    function mint(uint256 tokensToMint) public preExpiration() nonReentrant() returns (uint256 collateralUsed) {
+        collateralUsed = FixedPoint.fromUnscaledUint(tokensToMint).mul(collateralPerPair).rawValue;
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:96]: nit: should we be scaling tokensToMint? Seems like we should just cast it to a FixedPoint.Unsigned?

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+        require(longToken.burnFrom(msg.sender, longTokensToRedeem));
+        require(shortToken.burnFrom(msg.sender, shortTokensToRedeem));
+
+        // expirationTokensForCollateral is a number between 0 and 1e18. 0 means all collateral goes to short tokens and
+        // 1 means all collateral goes to the long token. Total collateral returned is the sum of payouts.
+        uint256 collateralPerToken = collateralPerPair.div(2);
+
+        uint256 longCollateralRedeemed =
+            FixedPoint
+                .fromUnscaledUint(longTokensToRedeem)
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:143]: Similar to above, I don't think we should be scaling token values, right?

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+                .mul(collateralPerToken)
+                .mul(expirationTokensForCollateral)
+                .rawValue;
+        uint256 shortCollateralRedeemed =
+            FixedPoint
+                .fromUnscaledUint(shortTokensToRedeem)
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:149]: Same here.

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        return IdentifierWhitelistInterface(finder.getImplementationAddress(OracleInterfaces.IdentifierWhitelist));
+    }
+
+    function _getOptimisticOracle() internal view returns (OptimisticOracleInterface) {
+        return OptimisticOracleInterface(finder.getImplementationAddress(OracleInterfaces.OptimisticOracle));
+    }
+
+    function _getAncillaryData() internal view returns (bytes memory) {
+        return abi.encodePacked(address(longToken), address(shortToken));
+    }
```
Comment (mrice32) [inline packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:182]: We don't need to do this in this PR, but do you think we should allow the deployer to customize the ancillary data?

**Review by mrice32 — COMMENTED** [review]
This looks awesome! A few minor comments. As mentioned offline, I think we can move forward with merging this pretty quickly (even without tests).

File: packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol (hunk 8 above)
Comment (chrismaree) [reply packages/core/contracts/financial-templates/contract-for-difference/ContractForDifference.sol:181]: sure. For now I left it like this just to follow the same pattern as the EMP. we totally should change this in the future.

**Review by mrice32 — APPROVED** [review]
LGTM!

**Pull request merged**